### ISSUE TITLE
[BUG][Fuzzer][wrong-code] `(x << k) >> k` on any sub-word int (`int8`/`uint8`/`int16`/`uint16`) silently returns the 32-bit-promoted value instead of the wrapped/sign-extended result

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2961]
### Required prerequisites

- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) and [Discussions](https://github.com/tile-ai/tilelang/discussions) that this hasn't already been reported. (comment there if it has.)
- [x] I have read the documentation <https://tilelang.com>.

### What version of TileLang are you using?

0.1.13

### System information

TileLang 0.1.13; NVIDIA L40S (sm_89), CUDA 12.8, PyTorch 2.8.

### Problem description

Extracting the low nibble of a `uint8` with `(b << 4) >> 4` (equivalently, sign-extending a packed sub-byte field of an `int8` with the same idiom) returns the wrong value on the scalar codegen path: the kernel compiles and runs, but every element is wrong. For `uint8` input `48`, `(b << 4) >> 4` yields **48** (the value is returned unchanged) where it should yield `48 & 0x0F == 0`; for `int8` input `8`, `(x << 4) >> 4` yields **8** where the sign-extended 4-bit field is `-8`. In every case the returned value is the input unchanged — the 32-bit `(x<<4)>>4` is the identity — instead of the narrow-width result.

The emitted CUDA is a bare C shift on the narrow operand:

```c
Out[...] = ((b << (uchar)4) >> (uchar)4);   // uchar b
Out[...] = ((x << (signed char)4) >> (signed char)4);   // signed char x
```

Under C integer promotion `uchar`/`signed char`/`short` operands widen to `int` before the shift, so `b << 4` is computed in 32 bits and is never truncated back to 8/16 bits. The `>> 4` then sees the un-truncated 32-bit intermediate. TileLang's TIR types the intermediate as the narrow dtype (so the language semantics are "wrap at 8/16 bits"), but the generated C does not enforce that width, so the two disagree. Same wrong result on `uint8` (240/256 elements wrong), `int8`, and `int16`. No traceback — the result is silently wrong and deterministic across runs.

<details><summary>Scope I checked — dtype × form grid (numerically ground-truthed), and the boundaries</summary>

Swept every integer width and both operand families through `(x<<k)>>k` on the scalar path (interleaved store), fresh cache, comparing against numpy computed **in the declared narrow dtype**; also checked float dtypes and the lone `x<<k` store. All on 0.1.13.

| dtype | `(x<<k)>>k` read-back | lone `x<<k` store |
|---|---|---|
| `int8` | **WRONG** (240/256; e.g. `(8<<4)>>4` → 8, want −8) | correct |
| `uint8` | **WRONG** (240/256; e.g. `(48<<4)>>4` → 48, want 0) | correct |
| `int16` | **WRONG** (k=4: 240/256; k=8: 254/256) | correct |
| `uint16` | **WRONG** (k=4: 240/256; k=8: 255/256) | correct |
| `int32` / `uint32` | correct (both k) — **upper bound** | correct |
| `float16`/`bfloat16`/`float32`/`float8` | **compile-time reject** (`Expected integer`) — **lower bound** | reject |

Boundaries this establishes:
- **Bounded to sub-32-bit integers.** `int32`/`uint32` are correct in both forms — the bug is exactly the promotion of a *narrow* C operand to 32-bit; at 32-bit there is no promotion. Every sub-word width (8 and 16, signed and unsigned) is wrong; the emitted operand cast differs (`signed char`/`short` vs `uchar`/`ushort`) but the missing truncation is identical.
- **Floats are cleanly rejected**, not silently wrong: shift on any float dtype fails the front-end `is_int()||is_uint()` guard at compile time. So the shift op *does* gate on dtype — it just gates on int-ness only, not on the sub-word *result width*, which is the gap here.
- **Only the read-back compose `(x<<k)>>k` trips it, never a lone `x<<k`.** Storing `x<<k` directly is correct for all narrow dtypes, because the store to the narrow buffer truncates the single result; it is the *intermediate* `(x<<k)` staying 32-bit and then feeding `>>k` un-truncated that loses the width.
- **Scalar path only.** The minimal repro forces it with an interleaved store (`Out[i*2+0]`); a plain `Out[i]` store gets vectorized and the vector path happens to compute the right value. Independent of whether the shift amount is a Python `int` or a `T.Cast`.

</details>

### Reproducible example code

```python
import tilelang, tilelang.language as T
import torch, numpy as np
N = 256

# BUG: extract the low nibble of a uint8 via (b << 4) >> 4.
# The interleaved store keeps this on the scalar (non-vectorized) codegen path.
@T.prim_func
def bug(A: T.Tensor((N,), "uint8"), Out: T.Tensor((N * 2,), "uint8")):
    with T.Kernel(1, threads=128):
        for i in T.Parallel(N):
            b = A[i]
            Out[i * 2 + 0] = (b << 4) >> 4   # want b & 0x0F  (uint8 wraps at 8 bits)
            Out[i * 2 + 1] = b >> 4          # high nibble (control op, always correct)

# CONTROL: same low nibble via a mask — no shift-left intermediate to truncate.
@T.prim_func
def ctrl(A: T.Tensor((N,), "uint8"), Out: T.Tensor((N * 2,), "uint8")):
    with T.Kernel(1, threads=128):
        for i in T.Parallel(N):
            b = A[i]
            Out[i * 2 + 0] = b & T.uint8(0x0F)
            Out[i * 2 + 1] = b >> 4

a  = torch.arange(N, dtype=torch.uint8).cuda()
an = a.cpu().numpy().astype(np.int64)
ref = np.zeros(N * 2, dtype=np.int64)
ref[0::2] = an & 0x0F          # numpy uint8 semantics
ref[1::2] = an >> 4

gb = tilelang.compile(bug,  out_idx=[1])(a).cpu().numpy().astype(np.int64)
gc = tilelang.compile(ctrl, out_idx=[1])(a).cpu().numpy().astype(np.int64)
print("BUG  nmis=%d got[32]=%d exp=%d ->" % (int((gb!=ref).sum()), gb[32], ref[32]),
      "FAIL" if not np.array_equal(gb, ref) else "PASS")   # FAIL: got 16, exp 0
print("CTRL nmis=%d got[32]=%d exp=%d ->" % (int((gc!=ref).sum()), gc[32], ref[32]),
      "PASS" if np.array_equal(gc, ref) else "FAIL")        # PASS
```

Output:

```
BUG  nmis=240 got[32]=16 exp=0 -> FAIL
CTRL nmis=0   got[32]=0  exp=0 -> PASS
```

### Traceback

No traceback — the kernel compiles and runs to completion; the result is silently wrong and deterministic.

### Expected behavior

`(x << k) >> k` should follow the dtype the operands are declared with, matching NumPy/C narrow-int semantics: `uint8` wraps at 8 bits (`48 → 0`), and the signed widths sign-extend the low field (`int8` `(8<<4)>>4 → -8`). This holds for every sub-word width tested — `int8`/`uint8`/`int16`/`uint16` — matching what `int32`/`uint32` already compute correctly and what the mask-based control (`b & 0x0F`) computes over the same inputs on the same path. The value is representable in the declared type; the shift form should produce it, not the 32-bit-promoted identity.

### Additional context

**Root cause.** The CUDA codegen emits a narrow-integer left shift as a bare C `<<` on the narrow operand and never truncates the result back to the type width, so C integer promotion computes the shift in `int` (32-bit) and the value the TIR types as `int8`/`uint8`/`int16` is never wrapped. The shift is printed verbatim at [`codegen_c.cc:671-674`](https://github.com/tile-ai/tvm/blob/0e15b274bce8b46f971abf5ac390e844aa6acee5/src/target/source/codegen_c.cc#L671-L674) (`PrintBinaryIntrinsic(op, " << ", …)`), with no cast to the sub-word result type — TileLang inherits this base-class emission. The consuming `>> 4` then reads the un-truncated 32-bit intermediate.

<details><summary>Why an explicit narrowing cast does not save it</summary>

Writing the intermediate as `T.Cast("int8", x << 4)` does not help: `x << 4` is already TIR-typed `int8`, so the cast is same-dtype and [`cast()` at `op.cc:433`](https://github.com/tile-ai/tvm/blob/0e15b274bce8b46f971abf5ac390e844aa6acee5/src/tir/op/op.cc#L433) returns the value unchanged (`if (value.dtype() == t) return value;`), and the RewriteSimplifier [`CastNode` rule](https://github.com/tile-ai/tvm/blob/0e15b274bce8b46f971abf5ac390e844aa6acee5/src/arith/rewrite_simplify.cc#L2494-L2497) drops it. That drop is consistent with TIR's wrap-at-type-width model; the mismatch is that codegen does not honor that model for the shift-left result. Materializing the intermediate through a narrow buffer (e.g. an `int8` fragment) does fix it, because the store truncates to 8 bits — confirming the truncation is what's missing:

```
signed char tmp[256];
tmp[tid] = (A[tid] << (signed char)4);   // stored -> truncated to int8
C[tid]   = (tmp[tid] >> (signed char)4); // correct
```
</details>

**Suggested fix.** One direction is for the codegen shift-left of a sub-word integer (`int8`/`uint8`/`int16`/`uint16`) to mask/cast the result back to the operand width, e.g. wrap the emitted `(a << b)` in a cast to the result dtype so the truncation the TIR type implies actually happens in C. Whether the right place is the codegen emission (`codegen_c.cc:671`) or an earlier lowering pass that inserts the narrowing for sub-word shifts is a call for the maintainers.

**Provenance.** The bare-`<<` emission at `codegen_c.cc:671` and the same-dtype `cast()` no-op are longstanding behaviors in the bundled TVM (`tile-ai/tvm` at the v0.1.9 submodule commit `0e15b274`); present at least since the codegen was inherited into TileLang. Introducing origin before that is unverified.

**Dedup.** I searched the open and closed tracker and found no existing report of this defect. The closed sub-word issues (#2477/#2479/#2482/#2483) are about the *value representation* of packed int2/int4 **decode** helpers (zero- vs sign-extension of a packed field), a different mechanism from an un-truncated shift intermediate.

**Reach.** The trigger is a narrow-integer left shift whose result is consumed before being stored/masked — a common kernel idiom and natural usage: `(x << k) >> k` is the standard way to sign-extend or extract a sub-byte field, and sub-word (`int8`/`uint8`/`int16`/`uint16`) shifts are pervasive in quantize/dequantize code. `grep` of the shipped `tilelang/quantize/` shows the shift idiom on narrow ints (`quantization.py:71-72` builds `((s << 8) | e_bf16) << 7` on `uint16`), but those sites `& mask` or `.astype(uint16)` immediately after the shift, which stores/truncates to the narrow width and dodges the bug; none use the read-back `(x<<k)>>k` field-extraction form. The vectorized codegen path also computes the right value, so a `Out[i] = (b<<4)>>4` (contiguous store, no interleave) does not trip it — the scalar path (interleaved/strided store, or any store the vectorizer declines) does. No test exercises the scalar sub-word shift-and-read-back path, so CI is green.

**Impact.** The trigger is narrow: a sub-word (`int8`/`uint8`/`int16`) `(x<<k)>>k` idiom that stays on the scalar codegen path (interleaved/strided store) and is consumed before any store or mask truncates it. When it fires there is no crash — the kernel compiles and runs, and every element of the affected region is silently wrong; the wrong value is deterministic but input-dependent (it tracks the un-truncated 32-bit intermediate), so a quantize/dequantize kernel using this field-extraction form would produce corrupt output that only differs from the reference on the high bits and could slip past casual checks. Fixing it closes the sub-word shift-width class so the emitted C honors the width the TIR already assigns, removing the codegen-vs-semantics disagreement.
